File: src/HER.m
function [ listSymbols ] = HER( image )
%HER The total process of applying the HandwrittenEquationRecognizer on an image

persistent latexLabel
if isempty(latexLabel)
    load('../data/extract/latexLabel.mat');
end

%angle = getRotationAngle(image);
%image = applyRotation(image,angle);
filteredImage = imbinarize(rgb2gray(image));
%filteredImage = filterImage(image);
resultImage = ones(size(filteredImage)+[20 20]);
filteredImage = double(repmat(filteredImage,1,1,3));
figure;
imshow(filteredImage);
listBB = getSymbolPositions(filteredImage);%,'maxclust',16);

for BB = listBB
    if((BB(1)-BB(2))*(BB(3)-BB(4))<10)
        continue
    end
    symbolImage = image(BB(1):BB(2),BB(3):BB(4),:);
    symbolImage = rgb2gray(symbolImage);
    symbolImage = double(symbolImage) - double(min(min(symbolImage)));
    symbolImage = double(symbolImage)/double(max(max(symbolImage)));
    assert(min(min(symbolImage))==0);
    assert(max(max(symbolImage))==1);
    sizeOfImage = max(size(symbolImage));
    addRows = sizeOfImage - size(symbolImage,1);
    addCols = sizeOfImage - size(symbolImage,2);
    symbolImage = [ ones(floor(addRows/2),size(symbolImage,2)) ; symbolImage];
    symbolImage = [ symbolImage; ones(ceil(addRows/2),size(symbolImage,2))];
    symbolImage = [ ones(size(symbolImage,1),floor(addCols/2)) , symbolImage];
    symbolImage = [ symbolImage, ones(size(symbolImage,1),ceil(addCols/2))];
    assert(size(symbolImage,1)==size(symbolImage,2));
    symbolImage = imresize(symbolImage, [32 32]);
    % tresh = prctile(symbolImage(:),16.5); %perform this over all symbols simultaneously
    % symbolImage = (symbolImage >= tresh);
    % symbolImage = imbinarize(symbolImage);
    prob = getRecognition(symbolImage);
    [bestp, besty] = max(prob);
    figure;
    imshow(symbolImage);
    title(sprintf('%f:%i\n%s',bestp,besty,latexLabel{besty}));
    lab = latexLabel{besty};
    oldlab = lab
    lab = ''
    for c=oldlab
        if c~='\'
            lab = [lab c];
        end
    end
    if lab(1)>='A' & lab(1)<='Z'
        lab(1) = char('a' + lab(1) - 'A');
    end
    if lab(1)=='|'
       lab = 'vert';
    end
    if lab(1)=='{'
       lab = 'lbrace';
    end
    if lab(1)=='-'
       lab = 'minus';
    end
    if lab(1)=='+'
       lab = 'plus';
    end
    if lab(1)=='$'
       lab = 'dollar';
    end
    if strcmp(lab,'mathscr{C}')
       lab = 'scrc';
    end
    if strcmp(lab,'mathcal{O}')
       lab = 'calo';
    end
    if strcmp(lab,'mathds{Z}')
       lab = 'bbz';
    end
    if strcmp(lab,'rightarrow')
       lab = 'longrightarrow';
    end
    if strcmp(lab,'mathbb{1}')
       lab = '1';
    end
    lab = [ '../data/symbols/' lab '.png' ]
    [labim, map, alpha] = imread(lab);
    [xs,ys] = size(labim);
    scale = ((BB(2)-BB(1))/xs + (BB(4)-BB(3))/ys)/2;
    labim = imresize(labim,scale);
    [xs,ys] = size(labim);
    mx = ceil((BB(1)+BB(2))/2);
    my = ceil((BB(3)+BB(4))/2);
    lx = mx - floor(xs/2);
    rx = mx + ceil(xs/2)-1;
    ly = my - floor(ys/2);
    ry = my + ceil(ys/2)-1;
    resultImage(lx+10:rx+10,ly+10:ry+10)=labim;
end

figure;
imshow(resultImage);

end

File: src/HER.py
import numpy as np
from predict import predict
import scipy.io
import matplotlib.pyplot as plt
import matlab.engine
eng = matlab.engine.start_matlab()

# load symbol data set
latexLabel = scipy.io.loadmat('../data/extract/latexLabel.mat')['latexLabel']
newLatexLabel = []
for i in range(len(latexLabel)):
    newLatexLabel.append(latexLabel[i][0][0])
latexLabel = newLatexLabel

#def rgb2gray(rgb):
#    return np.dot(rgb[...,:3], [0.299, 0.587, 0.114])

def HER(image):
    image = matlab.double(image.tolist())
    #angle = eng.getRotationAngle(image)
    #image = eng.applyRotation(image,angle)
    #plt.figure
    #plt.imshow(image)
    #plt.show()
    #filteredImage = eng.filterImage(image)
    #plt.figure
    #plt.imshow(filteredImage)
    #plt.show()
    listBB = eng.getSymbolPositions(image)

    for i in range(len(listBB[0])):
        lx = listBB[0][i]-1
        rx = listBB[1][i]
        ly = listBB[2][i]-1
        ry = listBB[3][i]
        print(lx,rx,ly,ry)
        symbolImage = matlab.uint8((np.array(image,dtype=np.uint8)[lx:rx,ly:ry,:]).tolist())
        symbolImage = np.array(eng.rgb2gray(symbolImage))
        (xs,ys) = symbolImage.shape
        if xs<ys:
            symbolImage = np.concatenate((np.ones((np.floor((ys-xs)/2),ys)),symbolImage,np.ones((np.ceil((ys-xs)/2),ys))), axis = 0)
        if xs>ys:
            symbolImage = np.concatenate((np.ones((xs,np.floor((xs-ys)/2))),symbolImage,np.ones((xs,np.ceil((xs-ys)/2)))), axis = 1)
        symbolImage = np.array(eng.imbinarize(eng.imresize(matlab.double(symbolImage.tolist()), matlab.int64([32,32]))))
        print(symbolImage.shape)
        result = predict(symbolImage)

        print(result)
        plt.figure
        plt.imshow(symbolImage)
        plt.title('prob:%.0f\tpred:%s'%(np.max(result)*100.,latexLabel[np.argmax(result)]))
        plt.show()
